### PR TITLE
#159204691: user should receive a json web token on authentication

### PR DESCRIPTION
**Pull request by chidioguejiofor**
#### What does this PR do?
- adds  signup test to the app
- adds a service folder to the app
- creates a .env-sample file to the app

#### Description of Task to be completed?
- as a user, I should receive a JSON Web Token on authentication

#### How should this be manually tested?
- run the command `npm install` to install all the dependencies
- run `npm start` in order to start the server
- create a post request to **localhost:3000/api/users/**. The body of this request should be similar to 
have the properties in the JSON below:
{
	"user" : {
		"email": "chidi@email.com",
		"username": "username101",
		"password": "password1"
	}
}
- you should receive a response like this:

{
    "data": {
        "email": "chidi@email.com",
        "username": "username101",
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImNoaWRpQGVtYWlsLmNvbSIsInVzZXJuYW1lIjoidXNlcm5hbWUxMDEiLCJwYXNzd29yZCI6InBhc3N3b3JkMSIsImlhdCI6MTUzMjQ1NDE0NCwiZXhwIjoxNTMyNzEzMzQ0fQ.njMCExUNOhkQwLk4vzE9ktCQvG1LBaWqNXfCnpprumw"
    }
}

#### Any background context you want to provide?
- Ensure that mongodb is set up and running
#### What are the relevant pivotal tracker stories?
- 159204691
#### Screenshots (if appropriate):
screen shot 2018-07-24 at 6 45 28 pm

#### Questions:
- None

### CODE REVIEW

File: .env-sample [hunk 1]
```diff
@@ -0,0 +1 @@
+JSON_SECRET=anysecretgoeshere
```
Comment (rhenshaw56) [inline .env-sample]: an appropriate name would be `AUTH_SECRET` or `JWT_SECRET` cos `JSON_SECRET` sounds abit confusing, don't you think ?

File: routes/api/users.js [hunk 2]
```diff
@@ -79,7 +81,17 @@ router.post("/users", function(req, res, next) {
 
     user.save()
         .then(function() {
-            return res.json({ user: user.toAuthJSON() });
+            console.log(TokenService);
```
Comment (rhenshaw56) [inline routes/api/users.js]: get rid of the `console` statements

File: routes/api/users.js [hunk 3]
```diff
@@ -79,7 +81,17 @@ router.post("/users", function(req, res, next) {
...
+            TokenService.createToken(req.body.user, '3d', function(token){
```
Comment (rhenshaw56) [inline routes/api/users.js]: first off! transform this to either return just the token and not make use of a callback cos then you're now forced to return a response in the context of the `TokenService.createToken`. What if you still had to do something else with the user data.

Also, define the token expiry inside where you're using it, cos it's redundant if you have to pass in a default expiry value to the function, and that just makes the function impure.

Finally, I'd like to hope that since you're creating the token with the whole `req.body.user` that the user's password ain't in there somehow

File: routes/api/users.js [hunk 4]
```diff
@@ -2,6 +2,8 @@ const mongoose = require("mongoose");
 const router = require("express").Router();
 const passport = require("passport");
 const User = mongoose.model("User");
+const TokenService = require('../../services/TokenService');
```
Comment (rhenshaw56) [inline routes/api/users.js]: this pascal case naming convention should only hold if `TokenService` is a class

File: test/signup.test.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        
+    });
+     describe('/api/users/', function(){
+        
+        describe('status code', function(){
+            it('should be equal to 201', function(done){
+                request.post('/api/users/')
+                .send({
+                    user: {
+                        email: 'goodemailformat@goodemail.com',
```
Comment (rhenshaw56) [inline test/signup.test.js]: there's this really nice npm module for having mock data called [faker](https://www.npmjs.com/package/faker) you could use that instead

File: test/signup.test.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+                .end(function(err, resp){
+                    expect(resp.body).property('data')
+                        .to.have.property('email');
+                        expect(resp.body).property('data')
+                        .to.have.property('username');
+                    done();
+                });
+            });
+        })
+     });
```
Comment (rhenshaw56) [inline test/signup.test.js]: don't you think it'd be nice to test when a user tries to signup without inputing any details or some invalid details or a user tries to signup but they were already members. What do you think ?

**Review by rhenshaw56 — CHANGES_REQUESTED** [review]

**Comment by rhenshaw56** [comment]
I see you integrated testing setup in these. could you quickly create a separate pr that shows the test setup so I could quickly get that in an then you can rebase this and update the tests

File: test/signup.test.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+            it('should be equal to 201', function(done){
+                request.post('/api/users/')
+                .send({
+                    user: {
+                        email: 'goodemailformat@goodemail.com',
+                        username: 'username101',
+                        password: 'passthat1password',
+                    }
+                })
+                .expect(201, done);
```
Comment (megame24) [inline test/signup.test.js]: I suggest you wrap your assertions with a `.end()` block, which should be independent of your `done` function this will ensure tests that deal with promises terminate properly.

File: services/TokenService.js [hunk 8]
```diff
@@ -0,0 +1,11 @@
+const jwt = require('jsonwebtoken');
+
+function createToken(payload,expiresIn, callback){
+    jwt.sign(payload,process.env.JSON_SECRET, {expiresIn}, (err, token)=> {
+        callback(token);
```
Comment (megame24) [inline services/TokenService.js]: I believe we should stick to promises in place of callbacks to ensure consistency. Also in this use case, you should just return the token without a callback.

File: test/signup.test.js (hunk 6 above)
Comment (chidioguejiofor) [reply test/signup.test.js]: I think that would be nice

Comment (chidioguejiofor) [reply test/signup.test.js]: I would add that

File: .env-sample (hunk 1 above)
Comment (chidioguejiofor) [reply .env-sample]: Thanks

File: routes/api/users.js (hunk 4 above)
Comment (chidioguejiofor) [reply routes/api/users.js]: Okay noted. I would see how to convert this to a class

Comment (chidioguejiofor) [reply routes/api/users.js]: I think I could convert this to a class once the refactor to es6 is done.

File: config/passport.js [hunk 9]
```diff
@@ -1,7 +1,5 @@
 const passport = require("passport");
 const LocalStrategy = require("passport-local").Strategy;
```
Comment (rhenshaw56) [inline config/passport.js]: change LocalStategy to camelCase

File: services/TokenService.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
+const jwt = require('jsonwebtoken');
+
+function createToken(payload,expiresIn, callback){
+    return jwt.sign(payload,process.env.AUTH_SECRET, {expiresIn});
```
Comment (rhenshaw56) [inline services/TokenService.js]: fix indentation

File: test/signup.test.js [hunk 11]
```diff
@@ -0,0 +1,59 @@
+const chai = require('chai');
+const app = require('../index');
+const request = require('supertest')(app);
+const db = require('../models/');
+const mocha = require('mocha');
+const faker = require('faker');
```
Comment (rhenshaw56) [inline test/signup.test.js]: nice!

File: test/signup.test.js [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+      db.User.destroy({
+          where: {},
+      }).then(user => console.log(user));
+      done();
+        
+    });
+     describe('/api/users/', function(){
+
+        describe('status code', function(){
+            it('should be equal to 201', function(done){
```
Comment (rhenshaw56) [inline test/signup.test.js]: a proper wording would be `should be equal to 201 when a user signs up successfully`

File: test/signup.test.js [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+                request.post('/api/users/')
+                .send({user: user})
+                .expect(201, done);
+            });
+        })
+        describe('request body', function(){
+            it('should have a data property that contains a token string property', function(done){
```
Comment (rhenshaw56) [inline test/signup.test.js]: `should have a token property`

File: test/signup.test.js [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+const user =  {
+    email: faker.internet.email(),
+    username:faker.name.lastName(),
+    password: faker.internet.password(),
+};
+
+
+describe('POST requests', function() {
+    beforeEach(function(done){
+      db.User.destroy({
```
Comment (rhenshaw56) [inline test/signup.test.js]: so since this feature flows through the the api endpoint not the db, a good approach would be hitting the signup endpoint with valid details for signup first and then do the your thing

**Review by rhenshaw56 — CHANGES_REQUESTED** [review]

**Comment by rhenshaw56** [comment]
you got a whole lot of indentation issues, you got to fix that mehn!

**Comment by chidioguejiofor** [comment]
I would correct those indentation issues once linting is set up. I cannot really see all the errors without linting

File: test/signup.test.js (hunk 14 above)
Comment (chidioguejiofor) [reply test/signup.test.js]: I don't understand what you mean

File: test/signup.test.js (hunk 13 above)
Comment (chidioguejiofor) [reply test/signup.test.js]: Do you mean that I should not put the token in the data property? What I have now is:
{
   data:{
     username: 'username',
    "token": "eycodac dscds",
    ...
  }
}

Comment (chidioguejiofor) [reply test/signup.test.js]: should I change it to

{
  data:{
         username: 'username',
         ...
   },
  "token": "eycodac dscds",
   .....
}

File: test/signup.test.js (hunk 12 above)
Comment (chidioguejiofor) [reply test/signup.test.js]: Noted

File: config/passport.js (hunk 9 above)
Comment (chidioguejiofor) [reply config/passport.js]: `LocalStrategy` is actually a class. 
If you look downwards it was instantiated when `passport.use` function was called

File: test/signup.test.js (hunk 13 above)
Comment (rhenshaw56) [reply test/signup.test.js]: change ~should have a data property that contains a token string~ to **should have a token property**

File: routes/api/users.js [hunk 15]
```diff
@@ -1,6 +1,9 @@
 const router = require("express").Router();
 const passport = require("passport");
 const User = require("../../models/").User;
+const TokenService = require('../../services/TokenService');
+const dotenv = require('dotenv/config');
+const AuthValidator = require('../../validators/AuthValidator');
```
Comment (rhenshaw56) [inline routes/api/users.js]: put the validator in a middlewares folder

File: test/signup.test.js [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+
+const describe = mocha.describe;
+
+const expect = chai.expect; 
+
+
+function generateUser(){
+    const user = {
+        email: faker.internet.email(),
+        username:faker.name.lastName() + Date.now(),
```
Comment (rhenshaw56) [inline test/signup.test.js]: instead if using `Date.now()` can you just trim out the email part from the `@` symbol and use that for the `username`

**Review by rhenshaw56 — CHANGES_REQUESTED** [review]

File: index.js [hunk 17]
```diff
@@ -12,6 +12,7 @@ const fs = require("fs"),
 require('dotenv').config();
 
 const isProduction = process.env.NODE_ENV === "production";
+const isTest = process.env.NODE_ENV === "test";
```
Comment (megame24) [inline index.js]: is the variable `isTest` necessary, I don't see anywhere you are making use of it.

File: test/signup.test.js [hunk 18]
```diff
@@ -0,0 +1,127 @@
+const chai = require('chai');
+const app = require('../index');
+const request = require('supertest')(app);
```
Comment (megame24) [inline test/signup.test.js]: I have set up the tests to use `chai-http`, can you check out the `index.test.js` file and follow the testing pattern used there so that everyone's test will be consistent going forward.

File: routes/api/users.js (hunk 15 above)
Comment (chidioguejiofor) [reply routes/api/users.js]: I actually synced with my teammates and found that someone else was working on validator so i removed it

File: .travis.yml [hunk 19]
```diff
@@ -1,5 +1,11 @@
 language: node_js
 
+
+before_script:
+  - npm install -g sequelize-cli
+  - psql -c 'create database AuthorsHavenTest;' -U postgres
+  - sequelize db:migrate --env test
```
Comment (megame24) [inline .travis.yml]: This code is not needed, it already exists in the `pretest` script.

File: config/passport.js [hunk 20]
```diff
@@ -10,14 +10,29 @@ passport.use(
...
+                   
+                    if (!user) {
                         return done(null, false, {
-                            errors: { "email or password": "is invalid" }
+                            errors: { "email": "is invalid" }
+                        });
+                    }
+                    if (!(hash ===  user.hash)) {
+                        return done(null, false, {
+                            errors: { "email": "is invalid" }
```
Comment (megame24) [inline config/passport.js]: The error message should be `{ "password": "is invalid" }` for this logic branch.

File: models/User.js [hunk 21]
```diff
@@ -8,18 +8,13 @@ module.exports = (sequelize, DataTypes) => {
       allowNull: false,
       unique: true,
       required: true,
-      validate: {
-        is: /^[a-zA-Z0-9]+$/i,
-      },
```
Comment (megame24) [inline models/User.js]: Schema validation has been removed, attend to it.

File: models/User.js [hunk 22]
```diff
@@ -8,18 +8,13 @@ module.exports = (sequelize, DataTypes) => {
...
     },
     email: {
       type: DataTypes.STRING,
       allowNull: false,
       unique: true,
       required: true,
-      validate: {
-        isEmail:true,
-      },
```
Comment (megame24) [inline models/User.js]: Schema validation has been removed, attend to it.

File: package.json [hunk 23]
```diff
@@ -6,12 +6,15 @@
   "scripts": {
     "start": "node index.js",
     "test": "cross-env NODE_ENV=test mocha ./test/**/*.js",
-    "coverage": "cross-env NODE_ENV=test ./node_modules/istanbul/lib/cli.js cover ./node_modules/mocha/bin/_mocha",
-    "lint": "eslint ."
```
Comment (mppdg) [inline package.json]: Why did you remove the `lint` script from `package.json`?

File: package.json [hunk 24]
```diff
@@ -6,12 +6,15 @@
...
+    "coverage": "npm run pretest && cross-env NODE_ENV=test ./node_modules/istanbul/lib/cli.js cover ./node_modules/mocha/bin/_mocha && npm run posttest",
+    "pretest": "sequelize --env test db:migrate && sequelize --env test db:seed",
```
Comment (megame24) [inline package.json]: You should undo migration before running `db:migrate` to ensure the tables are dropped before being recreated in a situation where your test ended with an error, and didn't get the chance to drop the table(`db:migrate:undo:all`). I suggest you edit the `pretest` script to look like this `npm run posttest && sequelize --env test db:migrate && sequelize --env test db:seed`

File: package.json [hunk 25]
```diff
@@ -6,12 +6,15 @@
...
+    "posttest": "sequelize --env test db:migrate:undo:all",
+    "migration": " sequelize db:migrate:undo:all && sequelize db:migrate   "
   },
```
Comment (tutugodfrey) [inline package.json]: You have removed the script for linting on line 10. Can you please add it so we can always lint to clear errors before pushing

File: config/passport.js [hunk 26]
```diff
@@ -10,14 +10,29 @@ passport.use(
             passwordField: "user[password]"
         },
         function(email, password, done) {
-            User.findOne({ email: email })
+            User
+            .findOne({
+                where: {
+                    email: email
+                }
+            })
```
Comment (tutugodfrey) [inline config/passport.js]: @chidioguejiofor can you use object destructing in the email field to avoid hound shouting.

``` 
where: {
  email,
}
```
will be ok

File: routes/api/users.js [hunk 27]
```diff
@@ -1,6 +1,8 @@
 const router = require("express").Router();
 const passport = require("passport");
 const User = require("../../models/").User;
+const TokenService = require('../../services/TokenService');
```
Comment (megame24) [inline routes/api/users.js]: Since `TokenService` is not a constructor, change the capital `T` to a small letter `t`.

File: test/auth.test.js [hunk 28]
```diff
@@ -0,0 +1,190 @@
...
+                .end((err, res)=>{
+                    expect(res.status).to.equal(200);
+                    expect(res.body).to.haveOwnProperty('data');
+                    expect(res.body.data).to.haveOwnProperty('token');
+                    done();
+                });
+            }); 
+    
+        });
+    });
```
Comment (megame24) [inline test/auth.test.js]: To fix your indentation issues, you can set your editor spaces to `2` at the bottom left corner of your VS code editor. Then press `option` + `shift` + `f` on the file to auto correct the indentation.

File: package.json [hunk 29]
```diff
@@ -33,6 +37,7 @@
     "request": "^2.87.0",
     "sequelize": "^4.38.0",
     "slug": "^0.9.1",
+    "supertest": "^3.1.0",
```
Comment (mppdg) [inline package.json]: You should remove the `supertest` dependency since you don't have the need for it.

File: test/auth.test.js [hunk 30]
```diff
@@ -0,0 +1,190 @@
...
+                });
+            }); 
+    
+        });
+    
+    
+        
+        describe('loging in with all the required fields', ()=>{
+            it('should return a 200 status code has a data property with a token', (done)=>{
+
```
[the remark this answers is not in the document]
Comment (tutugodfrey) [reply test/auth.test.js]: Please close all trailing the spaces for cleaner code

File: test/auth.test.js [hunk 31]
```diff
@@ -0,0 +1,190 @@
...
+                .post(LOGIN_ROUTE)
+                .send({
+                    user:{
+                        username: faker.name.lastName() + Date.now(),
+                    }
+                })
+                .end((err, res)=>{
+                    expect(res.body)
+                       .to.not.have.property('data');
+                    done();
```
Comment (tutugodfrey) [inline test/auth.test.js]: For every occurence of
```
.end((err, res)=>{
                    expect(res.body)
                       .to.not.have.property('data');
```
you should also check for the value in `data` object. That way you are sure the value is not the same with what is in

```
                .end((err, res)=>{
                    expect(res.status).to.equal(201 );
                    expect(res.body).to.haveOwnProperty('data');
                    expect(res.body.data).to.haveOwnProperty('token');
                    done();
                })
```

File: test/auth.test.js [hunk 32]
```diff
@@ -0,0 +1,190 @@
...
+        before((done)=>{
+            chai.request(server)
+            .post(SIGN_UP_ROUTE)
+            .send(tempUser)
+            .end((err, res)=>{
+               done();
+            });
+        });
+       
+        describe('sloging in with no username', ()=>{
```
Comment (JCanaks) [inline test/auth.test.js]: Correct typo "sloging in" to logging in

File: test/auth.test.js [hunk 33]
```diff
@@ -0,0 +1,190 @@
...
+const chaiHttp = require('chai-http');
+const server = require('../index');
+const supertest = require('supertest');
+const expect = chai.expect;
+const SIGN_UP_ROUTE = '/api/users';
+const LOGIN_ROUTE = '/api/users/login';
+const models = require('../models');
+chai.use(chaiHttp);
+
+function generateUser(){
```
Comment (JCanaks) [inline test/auth.test.js]: This function 'generateUser' is never used in the test. Suggest removing it

File: test/auth.test.js [hunk 34]
```diff
@@ -0,0 +1,170 @@
...
+    before((done) => {
+      chai.request(server)
+        .post(SIGN_UP_ROUTE)
+        .send(tempUser)
+        .end(() => {
+          done();
+        });
+    });
+
+    describe('sloging in with no username', () => {
```
Comment (JCanaks) [inline test/auth.test.js]: change 'sloging' to 'logging'

**Review by megame24 — APPROVED** [review]

**Review by JCanaks — APPROVED** [review]

**Review by mppdg — APPROVED** [review]

**Review by mppdg — APPROVED** [review]

File: .gitignore [hunk 35]
```diff
@@ -1,3 +1,4 @@
+package-lock.json
```
Comment (rhenshaw56) [inline .gitignore]: pls remove package-lock.json from here

File: index.js [hunk 36]
```diff
@@ -1,34 +1,33 @@
...
-app.use(require("method-override")());
-app.use(express.static(__dirname + "/public"));
+app.use(require('method-override')());
+
+app.use(express.static(`${__dirname}/public`));
 
 app.use(
   session({
-    secret: "authorshaven",
+    secret: 'authorshaven',
```
Comment (rhenshaw56) [inline index.js]: put this secret in a .env file

File: index.js [hunk 37]
```diff
@@ -39,25 +38,26 @@ if (!isProduction) {
...
+// / error handlers
 
 // development error handler
 // will print stacktrace
 if (!isProduction) {
-  app.use(function (err, req, res, next) {
+  app.use((err, req, res) => {
     // do not log errors on test environment, returning it is good enough
-    if(process.env.NODE_ENV !== 'test') console.log(err.stack);
+    if (process.env.NODE_ENV !== 'test') console.log(err.stack); // eslint-disable-line
```
Comment (rhenshaw56) [inline index.js]: remove the eslint disabling

File: routes/api/users.js [hunk 38]
```diff
@@ -1,87 +1,105 @@
...
-        });
-    }).catch(next);
+  User.create({ username, email, hash })
+    .then((user) => {
+      const payload = {
+        username: user.username,
+        email: user.email,
+      };
 
+      const token = tokenService.createToken(payload, '3d');
```
Comment (rhenshaw56) [inline routes/api/users.js]: I think I told you something about not directly passing `'3d'` into this call here, define it where you're using it

File: seeders/20180731154458-User.js [hunk 39]
```diff
@@ -0,0 +1,11 @@
+
+
+module.exports = {
+  up: queryInterface => queryInterface.bulkInsert('Users', [{
+    username: 'usename10101',
+    email: 'emadasfa21il@email.com',
+    password: 'passtheword'
```
Comment (rhenshaw56) [inline seeders/20180731154458-User.js]: read this details from `.env`

**Review by rhenshaw56 — CHANGES_REQUESTED** [review]

File: index.js [hunk 40]
```diff
@@ -10,6 +10,7 @@ import {} from './models/User';
 import {} from 'dotenv/config';
```
Comment (henshawrowland56) [inline index.js]: `import { config } from 'dotenv'`;
`import { passport } from './config'`
